File: q1/answer.js
//YTB Link: https://youtu.be/9UtInBqnCgA?si=g0ptibfe1ROdgW7h
//Answer 1
let nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10];

const nonDup = (numPass) => { //Time complexity O(n)
    let nonDuplicate = new Set(numPass);

    if (nonDuplicate.size === numPass.length) {
        return true;
    } else {
        return false;
    }
}

console.log(nonDup(nums));

//Answer 2

const nonDup2 = (numPass) => { //Time complexity O(n^2)
    for (let i=0; i<numPass.length; i++) {
        for (let j=i+1; j<numPass.length; j++) {
            if (numPass[i] === numPass[j]) {
                return false;
            }
        }
    }
    return true;
}

console.log(nonDup2(nums));

File: q3/answer.py
#YTB Link: https://youtu.be/KLlXCFG5TnA?si=IIgus7oN3W0CICZ9
from typing import List


class Solution:
    def twoSum( nums: List[int], target: int) -> List[int]:
        prevMap = {} #val : index
        for i, n in enumerate(nums): #enumerate gives us both index and value
            diff = target - n
            if diff in prevMap:
                return [prevMap[diff], i]
            prevMap[n] = i
        return []
    
print("two Sum: ", Solution.twoSum([2,7,11,15], 9))
# Time Complexity: O(n)
# Space Complexity: O(n)
# The time complexity of the twoSum function is O(n) because it iterates through the list of numbers once,
# and the space complexity is O(n) because it uses a dictionary to store the indices of the numbers.

